SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -64,69 +67,58 @@ void USpatialActorChannel::Init(UNetConnection* InConnection, int32 ChannelIndex
 {
 	Super::Init(InConnection, ChannelIndex, bOpenedLocally);
 
-	SpatialNetDriver = Cast<USpatialNetDriver>(Connection->Driver);
-	check(SpatialNetDriver);
-
-	WorkerView = SpatialNetDriver->GetSpatialOS()->GetView();
+	NetDriver = Cast<USpatialNetDriver>(Connection->Driver);
+	check(NetDriver);
+	Sender = NetDriver->Sender;
+	Receiver = NetDriver->Receiver;
 }
 
 void USpatialActorChannel::DeleteEntityIfAuthoritative()
 {
-	bool bHasAuthority = false;
-	USpatialInterop* Interop = SpatialNetDriver->GetSpatialInterop();
-
-	TSharedPtr<worker::View> PinnedView = WorkerView.Pin();
-	if (PinnedView.IsValid())
+	if (NetDriver->Connection == nullptr)
 	{
-		bHasAuthority = Interop->IsAuthoritativeDestructionAllowed()
-			&& PinnedView->GetAuthority<improbable::Position>(ActorEntityId.ToSpatialEntityId()) == worker::Authority::kAuthoritative;
+		return;
 	}
 
-	UE_LOG(LogSpatialGDKActorChannel, Log, TEXT("Delete Entity request on %d. Has authority: %d "), ActorEntityId.ToSpatialEntityId(), bHasAuthority);
+	bool bHasAuthority = NetDriver->IsAuthoritativeDestructionAllowed() && NetDriver->View->GetAuthority(EntityId, improbable::Position::ComponentId) == WORKER_AUTHORITY_AUTHORITATIVE;
+
+	UE_LOG(LogTemp, Log, TEXT("Delete entity request on %lld. Has authority: %d"), EntityId, (int)bHasAuthority);
 
 	// If we have authority and aren't trying to delete a critical entity, delete it
-	if (bHasAuthority && !IsCriticalEntity())
+	if (bHasAuthority && !IsSingletonEntity())
 	{
-		Interop->DeleteEntity(ActorEntityId);
+		Sender->SendDeleteEntityRequest(EntityId);
 	}
+
+	Receiver->CleanupDeletedEntity(EntityId);
 }
 
-bool USpatialActorChannel::IsCriticalEntity()
+bool USpatialActorChannel::IsSingletonEntity()
 {
-	// Don't delete if the actor is the spawner
-	if (ActorEntityId.ToSpatialEntityId() == SpatialConstants::EntityIds::SPAWNER_ENTITY_ID)
-	{
-		return true;
-	}
-
-	// Don't delete if the actor is a Singleton
-	PathNameToEntityIdMap* SingletonNameToEntityId = SpatialNetDriver->GetSpatialInterop()->GetSingletonNameToEntityId();
-
-	if (SingletonNameToEntityId == nullptr)
-	{
-		return false;
-	}
-
-	for(const auto& Pair : *SingletonNameToEntityId)
-	{
-		if (Pair.second == ActorEntityId.ToSpatialEntityId())
-		{
-			return true;
-		}
-	}
+	return NetDriver->GlobalStateManager->IsSingletonEntity(EntityId);
```

Comment:
Couldn't this be deduced from the actor so that we don't need the the dependency on the GSM and the Netdriver in this function? Will the actor always be valid on the actor channel?

Reply:
The Actor could be deleted by this point. We could store the Class*, but that seems like overkill.